File: tools/compare_all_ephemerides.py
#!/usr/bin/env python3
"""
Comprehensive ephemeris accuracy comparison.

Compares geocentric ecliptic longitude accuracy across multiple ephemerides:
- JPL DE440 (reference/gold standard)
- EPM2021 (Russian high-precision)
- Swiss Ephemeris (widely used in astrology)

Uses ANGULAR SEPARATION (arcseconds) as the primary metric - most meaningful
for practical astronomy/astrology applications.

Author: AI Assistant
Date: 2025-10-30
"""

import numpy as np
from pathlib import Path
import sys

try:
    import calceph
    HAS_CALCEPH = True
except ImportError:
    HAS_CALCEPH = False
    print("WARNING: calceph not found, DE440/EPM2021 comparison skipped")
    print("  Build calceph first: cd vendor/calceph-4.0.1 && cmake --build build")
    print()


class EphemerisComparison:
    """Compare ephemeris accuracy using geocentric ecliptic coordinates."""

    # Body IDs (NAIF)
    BODIES = {
        'Sun': 10,
        'Moon': 301,
        'Mercury': 1,
        'Venus': 2,
        'Mars': 4,
        'Jupiter': 5,
        'Saturn': 6,
        'Uranus': 7,
        'Neptune': 8,
        'Pluto': 9,
        'Earth': 399,
    }

    # Swiss Ephemeris body IDs (different!)
    SE_BODIES = {
        'Sun': 0,
        'Moon': 1,
        'Mercury': 2,
        'Venus': 3,
        'Mars': 4,
        'Jupiter': 5,
        'Saturn': 6,
        'Uranus': 7,
        'Neptune': 8,
        'Pluto': 9,
    }

    # Test epochs (covering different centuries)
    TEST_EPOCHS = [
        ('J1900.0', 2415020.0),
        ('J1950.0', 2433282.5),
        ('J2000.0', 2451545.0),
        ('2010-01-01', 2455197.5),
        ('2020-01-01', 2458849.5),
        ('2030-01-01', 2462502.5),
        ('2050-01-01', 2469807.5),
    ]

    def __init__(self):
        self.de440_file = Path("data/ephemerides/jpl/de440/linux_p1550p2650.440")
        self.epm2021_file = Path("data/ephemerides/epm/2021/epm2021.eph")
        self.ephe_path = Path("ephe")

        self.de440 = None
        self.epm2021 = None

    def load_ephemerides(self):
        """Load CALCEPH ephemerides."""
        if not HAS_CALCEPH:
            return False

        print("Loading ephemerides...")

        # Load JPL DE440
        if self.de440_file.exists():
            self.de440 = calceph.CalcephBin()
            self.de440.open(str(self.de440_file))
            print(f"  ✓ JPL DE440: {self.de440_file}")
        else:
            print(f"  ⚠ JPL DE440 not found: {self.de440_file}")

        # Load EPM2021
        if self.epm2021_file.exists():
            self.epm2021 = calceph.CalcephBin()
            self.epm2021.open(str(self.epm2021_file))
            print(f"  ✓ EPM2021: {self.epm2021_file}")
        else:
            print(f"  ⚠ EPM2021 not found: {self.epm2021_file}")

        print()
        return True

    def compute_geocentric_ecliptic(self, eph, body_id, jd):
        """
        Compute geocentric ecliptic longitude/latitude.

        Returns:
            tuple: (longitude_deg, latitude_deg, distance_au) or None
        """
        if eph is None:
            return None

        try:
            # Get heliocentric position of body
            pos_body = eph.compute(jd, 0.0, body_id, 0,
                                   calceph.Constants.UNIT_AU + calceph.Constants.USECSUN)

            # Get heliocentric position of Earth
            pos_earth = eph.compute(jd, 0.0, 399, 0,
                                    calceph.Constants.UNIT_AU + calceph.Constants.USECSUN)

            if pos_body is None or pos_earth is None:
                return None

            # Geocentric position
            dx = pos_body[0] - pos_earth[0]
            dy = pos_body[1] - pos_earth[1]
            dz = pos_body[2] - pos_earth[2]

            # Convert to ecliptic longitude/latitude
            # Ecliptic longitude: arctan2(y, x)
            lon_rad = np.arctan2(dy, dx)
            lon_deg = np.degrees(lon_rad) % 360.0

            # Distance
            r = np.sqrt(dx*dx + dy*dy + dz*dz)

            # Ecliptic latitude: arcsin(z/r)
            lat_rad = np.arcsin(dz / r)
            lat_deg = np.degrees(lat_rad)

            return (lon_deg, lat_deg, r)

        except Exception as e:
            print(f"    Error computing {body_id}: {e}")
            return None

    def angular_separation(self, lon1, lat1, lon2, lat2):
        """
        Calculate angular separation between two points on sphere.

        Uses haversine formula for accuracy.

        Returns:
            float: Angular separation in arcseconds
        """
        # Convert to radians
        lon1_rad = np.radians(lon1)
        lat1_rad = np.radians(lat1)
        lon2_rad = np.radians(lon2)
        lat2_rad = np.radians(lat2)

        # Haversine formula
        dlat = lat2_rad - lat1_rad
        dlon = lon2_rad - lon1_rad

        a = np.sin(dlat/2)**2 + np.cos(lat1_rad) * np.cos(lat2_rad) * np.sin(dlon/2)**2
        c = 2 * np.arcsin(np.sqrt(a))

        # Convert to arcseconds
        return np.degrees(c) * 3600.0

    def compare_calceph_sources(self):
        """Compare EPM2021 vs JPL DE440."""
        if not HAS_CALCEPH or self.de440 is None or self.epm2021 is None:
            return

        print("=" * 80)
        print("EPM2021 vs JPL DE440 Accuracy Comparison")
        print("=" * 80)
        print()
        print("Metric: Geocentric ecliptic angular separation (arcseconds)")
        print()

        # Results storage
        results = {body: [] for body in self.BODIES.keys() if body not in ['Earth']}

        for epoch_name, jd in self.TEST_EPOCHS:
            print(f"Epoch: {epoch_name} (JD {jd})")
            print("-" * 80)

            for body_name, body_id in self.BODIES.items():
                if body_name == 'Earth':
                    continue

                # Compute from both sources
                de440_pos = self.compute_geocentric_ecliptic(self.de440, body_id, jd)
                epm_pos = self.compute_geocentric_ecliptic(self.epm2021, body_id, jd)

                if de440_pos is None or epm_pos is None:
                    print(f"  {body_name:10s}: Data unavailable")
                    continue

                lon_de, lat_de, r_de = de440_pos
                lon_epm, lat_epm, r_epm = epm_pos

                # Angular separation
                sep_arcsec = self.angular_separation(lon_de, lat_de, lon_epm, lat_epm)

                # Longitude difference (for debugging)
                dlon = abs(lon_epm - lon_de)
                if dlon > 180:
                    dlon = 360 - dlon
                dlon_arcsec = dlon * 3600

                results[body_name].append(sep_arcsec)

                # Format output
                if sep_arcsec < 1.0:
                    status = "✅"
                elif sep_arcsec < 60.0:
                    status = "⚠️"
                else:
                    status = "❌"

                print(f"  {body_name:10s}: {sep_arcsec:8.3f}\" {status}  " +
                      f"(lon: {lon_epm:7.3f}° vs {lon_de:7.3f}°, Δ={dlon_arcsec:.1f}\")")

            print()

        # Summary statistics
        print("=" * 80)
        print("SUMMARY: EPM2021 Accuracy vs JPL DE440")
        print("=" * 80)
        print()
        print(f"{'Body':<12} {'Samples':>8} {'Mean':>10} {'Median':>10} {'Max':>10} {'Status'}")
        print("-" * 80)

        for body_name in ['Sun', 'Moon', 'Mercury', 'Venus', 'Mars',
                          'Jupiter', 'Saturn', 'Uranus', 'Neptune', 'Pluto']:
            if body_name not in results or len(results[body_name]) == 0:
                continue

            errors = np.array(results[body_name])
            mean_err = np.mean(errors)
            median_err = np.median(errors)
            max_err = np.max(errors)

            # Status
            if median_err < 1.0:
                status = "✅ Excellent"
            elif median_err < 10.0:
                status = "✅ Good"
            elif median_err < 60.0:
                status = "⚠️ Acceptable"
            else:
                status = "❌ Poor"

            print(f"{body_name:<12} {len(errors):>8} " +
                  f"{mean_err:>9.2f}\" {median_err:>9.2f}\" {max_err:>9.2f}\" {status}")

        print()

    def compare_with_swisseph(self):
        """Compare Swiss Ephemeris vs JPL DE440 using FFI."""
        print("=" * 80)
        print("Swiss Ephemeris vs JPL DE440 Accuracy Comparison")
        print("=" * 80)
        print()
        print("Implementation: PHP FFI (run separate script)")
        print("Script: php/examples/compare_all_ephemerides.php")
        print()
        print("This will be implemented in PHP due to better Swiss Eph integration.")
        print()

    def run(self):
        """Run comparison."""
        print("=" * 80)
        print("Comprehensive Ephemeris Accuracy Comparison")
        print("=" * 80)
        print()

        if self.load_ephemerides():
            self.compare_calceph_sources()

        self.compare_with_swisseph()

        # Cleanup
        if self.de440:
            self.de440.close()
        if self.epm2021:
            self.epm2021.close()


def main():
    """Main entry point."""
    comparison = EphemerisComparison()
    comparison.run()

    print("=" * 80)
    print("COMPARISON COMPLETE")
    print("=" * 80)
    print()
    print("Note: For Swiss Ephemeris comparison, run:")
    print("  php php/examples/compare_all_ephemerides.php")
    print()


if __name__ == '__main__':
    main()

File: php/examples/compare_all_ephemerides.php
<?php
/**
 * Comprehensive ephemeris accuracy comparison.
 *
 * Compares geocentric ecliptic longitude accuracy across:
 * - JPL DE440 (NASA, reference/gold standard) - via Swiss Ephemeris
 * - EPM2021 (IAA RAS, Russian high-precision)
 * - Swiss Ephemeris (based on JPL DE440/DE431)
 *
 * NOTE: Swiss Ephemeris USES JPL DE440 data internally, making this
 * effectively a direct comparison: EPM2021 vs JPL DE440.
 *
 * Uses ANGULAR SEPARATION (arcseconds) as primary metric.
 *
 * Author: AI Assistant
 * Date: 2025-10-30
 */

// Load Swiss Ephemeris via FFI
$swissephDll = __DIR__ . '/../../vendor/swisseph/swedll64.dll';
if (!file_exists($swissephDll)) {
    die("ERROR: Swiss Ephemeris DLL not found: $swissephDll\n");
}

$ffi = FFI::cdef("
    void swe_set_ephe_path(char *path);
    int swe_calc_ut(double tjd_ut, int ipl, int iflag, double *xx, char *serr);
    int swe_calc(double tjd_et, int ipl, int iflag, double *xx, char *serr);
    double swe_deltat(double tjd_ut);
    void swe_close(void);
", $swissephDll);

// Set ephemeris path
$ephePath = __DIR__ . '/../../ephe';
$ffi->swe_set_ephe_path($ephePath);

// Swiss Ephemeris constants
const SE_SUN = 0;
const SE_MOON = 1;
const SE_MERCURY = 2;
const SE_VENUS = 3;
const SE_MARS = 4;
const SE_JUPITER = 5;
const SE_SATURN = 6;
const SE_URANUS = 7;
const SE_NEPTUNE = 8;
const SE_PLUTO = 9;

const SEFLG_SWIEPH = 2;      // Use Swiss Ephemeris files
const SEFLG_SPEED = 256;     // Calculate velocity

// Body mapping
$bodies = [
    'Sun' => SE_SUN,
    'Moon' => SE_MOON,
    'Mercury' => SE_MERCURY,
    'Venus' => SE_VENUS,
    'Mars' => SE_MARS,
    'Jupiter' => SE_JUPITER,
    'Saturn' => SE_SATURN,
    'Uranus' => SE_URANUS,
    'Neptune' => SE_NEPTUNE,
    'Pluto' => SE_PLUTO,
];

// Test epochs
$testEpochs = [
    ['J1900.0', 2415020.0],
    ['J1950.0', 2433282.5],
    ['J2000.0', 2451545.0],
    ['2010-01-01', 2455197.5],
    ['2020-01-01', 2458849.5],
    ['2030-01-01', 2462502.5],
    ['2050-01-01', 2469807.5],
];

// Load autoloader
require_once __DIR__ . '/../../vendor/autoload.php';

use Swisseph\Ephemeris\EphReader;
use Swisseph\Ephemeris\TimeScaleConverter;

$de440File = __DIR__ . '/../../data/ephemerides/jpl/de440/de440.eph';
$epm2021File = __DIR__ . '/../../data/ephemerides/epm/2021/epm2021.eph';

$hasDE440 = false;
$hasEPM2021 = false;
$hasSwissEph = true;  // FFI loaded above

echo "Loading ephemerides...\n";

if (file_exists($de440File)) {
    try {
        $de440 = new EphReader($de440File);
        $hasDE440 = true;
        $size = round(filesize($de440File) / 1024 / 1024, 1);
        echo "✓ Loaded JPL DE440 (DIRECT): {$size} MB\n";
    } catch (Exception $e) {
        echo "⚠ Failed to load DE440: " . $e->getMessage() . "\n";
    }
} else {
    echo "⚠ JPL DE440 not found: $de440File\n";
    echo "  Run: python tools/spice2eph.py data/ephemerides/jpl/de440/de440.bsp data/ephemerides/jpl/de440/de440.eph\n";
}

if (file_exists($epm2021File)) {
    try {
        $epm2021 = new EphReader($epm2021File);
        $hasEPM2021 = true;
        echo "✓ Loaded EPM2021: $epm2021File\n";
    } catch (Exception $e) {
        echo "⚠ Failed to load EPM2021: " . $e->getMessage() . "\n";
    }
}

echo "\n";

/**
 * Compute geocentric ecliptic coordinates from Swiss Ephemeris.
 *
 * NOTE: JPL ephemerides use TDB (Barycentric Dynamical Time).
 * Swiss Eph swe_calc() expects ET (Ephemeris Time ≈ TDB for our purposes).
 * We use swe_calc() directly with JD assuming it's already in TDB/ET.
 */
function computeSwissEph($ffi, int $bodyId, float $jd): ?array
{
    $xx = FFI::new("double[6]");
    $serr = FFI::new("char[256]");

    // Calculate geocentric ecliptic coordinates
    // Use swe_calc() with TDB/ET time (NOT swe_calc_ut() which expects UT)
    // Default flags return ecliptic lon/lat/dist
    $ret = $ffi->swe_calc($jd, $bodyId, SEFLG_SWIEPH | SEFLG_SPEED,
                          FFI::addr($xx[0]), FFI::addr($serr[0]));

    if ($ret < 0) {
        $error = FFI::string($serr);
        echo "    Swiss Eph error: $error\n";
        return null;
    }

    return [
        'lon' => $xx[0],  // Ecliptic longitude (degrees)
        'lat' => $xx[1],  // Ecliptic latitude (degrees)
        'dist' => $xx[2], // Distance (AU)
        'dlon' => $xx[3], // Longitude velocity (deg/day)
    ];
}

/**
 * Compute geocentric ecliptic coordinates from CALCEPH ephemeris.
 */
function computeFromEph(EphReader $eph, int $naifId, float $jd): ?array
{
    try {
        // Special handling for Sun and Moon
        if ($naifId === 10) {
            // Sun: invert Earth position to get geocentric Sun
            $posEarth = $eph->compute(399, $jd);
            if (!$posEarth) return null;

            $dx = -$posEarth['pos'][0];
            $dy = -$posEarth['pos'][1];
            $dz = -$posEarth['pos'][2];
        } elseif ($naifId === 301) {
            // Moon: geocentric position is Moon - Earth
            $posMoon = $eph->compute(301, $jd);
            $posEarth = $eph->compute(399, $jd);

            if (!$posMoon || !$posEarth) return null;

            $dx = $posMoon['pos'][0] - $posEarth['pos'][0];
            $dy = $posMoon['pos'][1] - $posEarth['pos'][1];
            $dz = $posMoon['pos'][2] - $posEarth['pos'][2];
        } else {
            // Planets: geocentric = heliocentric_planet - heliocentric_earth
            $posBody = $eph->compute($naifId, $jd);
            $posEarth = $eph->compute(399, $jd);

            if (!$posBody || !$posEarth) return null;

            $dx = $posBody['pos'][0] - $posEarth['pos'][0];
            $dy = $posBody['pos'][1] - $posEarth['pos'][1];
            $dz = $posBody['pos'][2] - $posEarth['pos'][2];
        }

        // Convert Cartesian equatorial (ICRF) to ecliptic coordinates
        // Apply obliquity of ecliptic (J2000): ε = 23.43929111°
        $epsilon = deg2rad(23.43929111);

        // Rotation matrix from equatorial to ecliptic:
        // x_ecl = x_eq
        // y_ecl = y_eq * cos(ε) + z_eq * sin(ε)
        // z_ecl = -y_eq * sin(ε) + z_eq * cos(ε)
        $x_ecl = $dx;
        $y_ecl = $dy * cos($epsilon) + $dz * sin($epsilon);
        $z_ecl = -$dy * sin($epsilon) + $dz * cos($epsilon);

        // Convert to spherical ecliptic lon/lat
        $lon = rad2deg(atan2($y_ecl, $x_ecl));
        if ($lon < 0) $lon += 360.0;

        $r = sqrt($x_ecl*$x_ecl + $y_ecl*$y_ecl + $z_ecl*$z_ecl);
        $lat = rad2deg(asin($z_ecl / $r));        return [
            'lon' => $lon,
            'lat' => $lat,
            'dist' => $r,
        ];

    } catch (Exception $e) {
        return null;
    }
}/**
 * Calculate angular separation using haversine formula.
 */
function angularSeparation(float $lon1, float $lat1, float $lon2, float $lat2): float
{
    $lon1Rad = deg2rad($lon1);
    $lat1Rad = deg2rad($lat1);
    $lon2Rad = deg2rad($lon2);
    $lat2Rad = deg2rad($lat2);

    $dlat = $lat2Rad - $lat1Rad;
    $dlon = $lon2Rad - $lon1Rad;

    $a = sin($dlat/2) ** 2 + cos($lat1Rad) * cos($lat2Rad) * sin($dlon/2) ** 2;
    $c = 2 * asin(sqrt($a));

    // Convert to arcseconds
    return rad2deg($c) * 3600.0;
}

/**
 * Format angular separation with status.
 */
function formatSeparation(float $arcsec): string
{
    if ($arcsec < 1.0) {
        $status = "✅";
    } elseif ($arcsec < 60.0) {
        $status = "⚠️";
    } else {
        $status = "❌";
    }

    return sprintf("%8.3f\" %s", $arcsec, $status);
}

// NAIF ID mapping
$naifIds = [
    'Sun' => 10,
    'Moon' => 301,
    'Mercury' => 1,
    'Venus' => 2,
    'Mars' => 4,
    'Jupiter' => 5,
    'Saturn' => 6,
    'Uranus' => 7,
    'Neptune' => 8,
    'Pluto' => 9,
];

// Results storage
$resultsSwissVsDE440 = [];
$resultsEPMVsDE440 = [];

echo str_repeat("=", 80) . "\n";
echo "Comprehensive Ephemeris Accuracy Comparison\n";
echo str_repeat("=", 80) . "\n\n";
echo "Metric: Geocentric ecliptic angular separation (arcseconds)\n";
echo "Reference: JPL DE440 (DIRECT from .eph file)\n\n";

foreach ($testEpochs as [$epochName, $jd]) {
    echo "Epoch: $epochName (JD $jd)\n";
    echo str_repeat("-", 80) . "\n";

    foreach ($bodies as $bodyName => $seId) {
        $naifId = $naifIds[$bodyName];

        // Compute DE440 position (REFERENCE)
        $de440Pos = null;
        if ($hasDE440) {
            $de440Pos = computeFromEph($de440, $naifId, $jd);
        }

        if (!$de440Pos) {
            echo sprintf("  %-10s: DE440 unavailable\n", $bodyName);
            continue;
        }

        // Compare EPM2021 vs DE440
        $epmSep = null;
        if ($hasEPM2021) {
            $epmPos = computeFromEph($epm2021, $naifId, $jd);
            if ($epmPos) {
                $epmSep = angularSeparation(
                    $de440Pos['lon'], $de440Pos['lat'],
                    $epmPos['lon'], $epmPos['lat']
                );

                if (!isset($resultsEPMVsDE440[$bodyName])) {
                    $resultsEPMVsDE440[$bodyName] = [];
                }
                $resultsEPMVsDE440[$bodyName][] = $epmSep;
            }
        }

        // Compare Swiss Eph vs DE440
        $swissSep = null;
        if ($hasSwissEph) {
            $swissPos = computeSwissEph($ffi, $seId, $jd);
            if ($swissPos) {
                $swissSep = angularSeparation(
                    $de440Pos['lon'], $de440Pos['lat'],
                    $swissPos['lon'], $swissPos['lat']
                );

                if (!isset($resultsSwissVsDE440[$bodyName])) {
                    $resultsSwissVsDE440[$bodyName] = [];
                }
                $resultsSwissVsDE440[$bodyName][] = $swissSep;
            }
        }

        // Display results
        $line = sprintf("  %-10s: DE440=%7.3f°", $bodyName, $de440Pos['lon']);

        if ($epmSep !== null) {
            $line .= sprintf(" | EPM Δ=%s", formatSeparation($epmSep));
        }

        if ($swissSep !== null) {
            $line .= sprintf(" | Swiss Δ=%s", formatSeparation($swissSep));
        }

        echo "$line\n";
    }

    echo "\n";
}

// Summary statistics
function printSummaryTable(array $results, string $title): void
{
    echo str_repeat("=", 80) . "\n";
    echo "$title\n";
    echo str_repeat("=", 80) . "\n\n";

    if (empty($results)) {
        echo "No data available\n\n";
        return;
    }

    printf("%-12s %8s %10s %10s %10s %s\n",
           'Body', 'Samples', 'Mean', 'Median', 'Max', 'Status');
    echo str_repeat("-", 80) . "\n";

    foreach (['Sun', 'Moon', 'Mercury', 'Venus', 'Mars', 'Jupiter', 'Saturn', 'Uranus', 'Neptune', 'Pluto'] as $bodyName) {
        if (!isset($results[$bodyName]) || empty($results[$bodyName])) {
            continue;
        }

        $errors = $results[$bodyName];
        $mean = array_sum($errors) / count($errors);

        sort($errors);
        $median = $errors[intval(count($errors) / 2)];
        $max = max($errors);

        // Status
        if ($median < 1.0) {
            $status = "✅ Excellent";
        } elseif ($median < 10.0) {
            $status = "✅ Good";
        } elseif ($median < 60.0) {
            $status = "⚠️ Acceptable";
        } else {
            $status = "❌ Poor";
        }

        printf("%-12s %8d %9.2f\" %9.2f\" %9.2f\" %s\n",
               $bodyName, count($errors), $mean, $median, $max, $status);
    }

    echo "\n";
}

// Print comparison summaries
printSummaryTable($resultsEPMVsDE440, "SUMMARY 1: EPM2021 vs JPL DE440 (DIRECT)");
printSummaryTable($resultsSwissVsDE440, "SUMMARY 2: Swiss Ephemeris vs JPL DE440 (Validation)");

echo str_repeat("=", 80) . "\n";
echo "NOTES\n";
echo str_repeat("=", 80) . "\n";
echo "• JPL DE440 used as reference (NASA gold standard)\n";
echo "• Angular separation measures actual sky position difference\n";
echo "• Arcseconds: < 1\" = Excellent, < 60\" = Good, > 60\" = Poor\n";
echo "• For comparison: Moon diameter ≈ 1800\", Jupiter ≈ 40\"\n";
echo "• Swiss Eph should match DE440 within < 0.1\" (uses same data)\n";
echo "\n";

// Cleanup
$ffi->swe_close();

echo "COMPARISON COMPLETE\n";
